lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -2,118 +2,333 @@ define((require, exports, module) ->
+      if @contextPreviewExists
+        [title, rendered] = @getCurrentFormula()
+        @popoverHandler.setContent(@getFormulaElement(), title, rendered)
+
+    updateCallback: =>
+      if @lastChangeTime?
+        curTime = Date.now()
+        @currentDelayedUpdateId = setTimeout(@updateCallback, ContextHandler.UPDATE_DELAY - (Date.now() - @lastChangeTime))
+        @lastChangeTime = null
+      else
+        @currentDelayedUpdateId = null
+        if @contextPreviewExists
+          curContext = LatexParsingContext.getContext(@editor.getSession(), @editor.getCursorPosition().row)
+          if curContext != "equation"
+            @destroyContextPreview()
           else
-            curSequenceIndex = 0
-            curEquationStart = null
-        return curEquationStart
-
-      getEquationRange: (row, column) ->
-        tokenIterator = new TokenIterator(editor.getSession(), row, column)
-        end = erh.getEquationEnd(tokenIterator)
-        start = erh.getEquationStart(tokenIterator)
-        return new Range(start.row, start.column, end.row, end.column)
-    }
-    return erh
+            @updateRange()
+            @updatePopover()
+
+    delayedUpdatePopover: =>
+      curDocLength = @editor.getSession().getLength()
+      if curDocLength != @prevDocLength
+        setTimeout(@updatePosition, 0)
+        @prevDocLength = curDocLength
+
+      if @currentDelayedUpdateId?
+        @lastChangeTime = Date.now()
+        return
+
+      @currentDelayedUpdateId = setTimeout(@updateCallback, ContextHandler.UPDATE_DELAY)
+
+    createContextPreview: ->
+      @updateRange()
+      @contextPreviewExists = true
+      if not katex?
+        initKaTeX(@initPopover)
+      else
+        @initPopover()
+      @prevDocLength = @editor.getSession().getLength()
+      @editor.on("change", @delayedUpdatePopover)
+      @editor.getSession().on("changeScrollTop", @updatePosition)
+
+    destroyContextPreview: ->
+      @curRange = null
+      @contextPreviewExists = false
+      @editor.off("change", @delayedUpdatePopover)
+      @editor.getSession().off("changeScrollTop", @updatePosition)
+      @popoverHandler.destroy(@getFormulaElement())
+
+    handleCurrentContext: => setTimeout((=>
+      if @currentDelayedUpdateId?
+        return
 
-  exports.getEquationRangeHandler = getEquationRangeHandler
+      {row: cursorRow, column: cursorColumn} = @editor.getCursorPosition()
+      currentContext = LatexParsingContext.getContext(@editor.getSession(), cursorRow)
+
+      if @contextPreviewExists and currentContext != "equation"
+        @destroyContextPreview()
+
+      else if not @contextPreviewExists and currentContext == "equation"
+        @createContextPreview()
+    ), 0)
+
+
+  class ConstrainedTokenIterator
+    constructor: (@session, @range, row, column) ->
+      @tokenIterator = new TokenIterator(@session, row, column)
+      curToken = @tokenIterator.getCurrentToken()
+      if not curToken?
+        @expired = false
```

Comment:
Normally "expired" is used when talking about time. Consider "outOfRange" or "isOut" or "isStopped/Terminated"